### DISCUSSION TITLE
Support for other file types

### DISCUSSION THREAD

**Discussion by chrisgurney** [chrisgurney/obsidian-note-toolbar#104] [Ideas]
I've been thinking about how toolbars should work for other filetypes, that aren't notes, including canvases and home screens.

The good news is that I have toolbars showing on the canvas locally. But, there's still some thinking to do, so I'm reluctant to release as-is.

## Example: How it might work for canvases

- The `top` and `floating button` positions are supported.
  - Toolbars in the `props` position are automatically changed to the `top` position on render (rather than not showing them at all).
- Right now canvas toolbars are applied based on **folder mappings**.

## Feedback wanted

_Would appreciate thoughts on how you would like to use toolbars on canvases, and other filet types!_

- My concern is that just enabling this would lead to toolbars taking up too much screen real-estate (particularly on mobile), without providing the means to control other filetype-specific toolbars. (e.g., for canvases, there's no properties, in particular.)
- I suspect the need will be to map toolbars better suited for working in these filetypes.
- All of the above would require a more granular method of mapping toolbars.
  - The `File Pattern` mapping type that I have planned could be one solution, that would maintain flexibility with other potential future file types: e.g., mapping to `*.canvas`

## Canvas demo

(To fix: Left-aligned items don't completely left-align to the screen in the `top` position.)

![CleanShot 2024-07-23 at 12 55 56@2x](https://github.com/user-attachments/assets/a4af3bf7-8f59-494f-8505-b3d114702a40)

**Comment by chrisgurney** [comment]
From u/ontorealist on [Reddit](https://www.reddit.com/r/ObsidianMD/comments/1d9idlj/comment/lf3h45w/):

> I don’t have the expertise to offer much, but from what I understand, I like that canvas toolbars map onto folders as well that works for my use-case and I don’t see it an inconvenience that it is not mapped onto the .canvas file itself currently?

> I’m not currently aware of or using any open source .canvas file if that’s what the concern is?

**Comment by chrisgurney** [comment]
Related: https://github.com/chrisgurney/obsidian-note-toolbar/discussions/130

**Comment by azaol-aegnor** [comment]
*About the "more granular method of mapping"*

An (additional?) option to map a toolbar to non-Markdown views/file could be to pull a list from `app.viewRegistry` from the configuration screen.

A "picture" view could be bound to png and jpg files and allow for a simpler configuration for the toolbar bindings, specifying a given toolbar should be available of "picture" views instead of a regez that catches every possible image extensions.

**Reply by azaol-aegnor** [reply]
I believe there's a view type registered for Canva files, *to be confirmed*

**Reply by chrisgurney** [reply]
I was looking at using the `data-type` attribute here
`<div class="workspace-leaf-content" data-type="canvas">`

**Reply by chrisgurney** [reply]
@azaol-aegnor Oh I meant to thank you for the pointer to `app.viewRegistry`! Thanks!

**Reply by chrisgurney** [reply]
@azaol-aegnor Follow up: I installed the [Kanban](https://github.com/mgmeyers/obsidian-kanban) plugin, and I do see `data-type="kanban"` on that element. 

I don't see `kanban` in the `app.viewRegistry.typeByExtension` list, but it _is_ in `app.viewRegistry.viewByType`. (Kanban files are created as markdown files, so I think that makes sense.)

I'm wondering out loud, I guess, what list specifically to offer to users in the settings UI.

**Reply by azaol-aegnor** [reply]
I'm not sure if it's normal or not. We need a better expert of the Obsidian API to know =)

**Comment by azaol-aegnor** [comment]
*Related to the topic of the discussion*

I don't use Canvas much. In my experience/usage of it, I would only have like one button on a canva. Therefor, I'd somewhat like to see it added to the sidebar rather than as a banner at the top.

I'd like to eat about other people use cases as well 😇

**Reply by chrisgurney** [reply]
The plan is to support both the `Top` position of the toolbar, and the `Floating button` as well.

**Reply by azaol-aegnor** [reply]
I meant the "native" right side toolbar with the lock and magnify buttons. As end user, it looks good. As a developper... It don't think it does.

Top bar may work in my opinion!
